stm32: mco: Remove deprecated Kconfig configuration

FRASTM opened the pull request:
change the configuration of the MCO 1 or 2 output for the stm32 series
Kconfig is no more used, set the configuration by the DTS  as speciied by the  dts/bindings/clock/st,stm32-clock-mco.yaml

See the refman of the stm32 to put the correct clock source and the prescaler value

```
&mco2 {
	clocks = <&rcc STM32_SRC_PLLI2S_R MCO2_SEL(1)>;
	prescaler = <MCO2_PRE(5)>; /* div by 5 --> set 0b111 in the MCO2PRE of the RCC CFGR */
	pinctrl-0 = <&rcc_mco_2_pc9>;
	pinctrl-names = "default";
	status = "okay";
};

```
Fixes https://github.com/zephyrproject-rtos/zephyr/issues/80885

FRASTM committed:
dts: arm: stm32f411 compatible for PLL I2S

The stm32f411 and stm32f412 and stm32f446 have a PLLI2S
with a div M in front of the PLLI2S input.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: clock_control: stm32 mco driver get define from DTS

Rely on the DTS to get the MCO input source clock and prescaler.
DTS configuration has been introduced and Kconfig method
deprecated two releases before and can be then safely removed.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
samples: boards: stm32 mco output control

This sample enables and configures the MCO1/2 output
for stm32 target boards.
MCO_PRE_DIV_n is defined by the stm32XX_clock.h

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
include: bindings: high speed clock for the stm32F1/F4 series

The stm32F4 serie has HSI and HSE clock possible clock sources
The stm32F1 also has an external HSE possible clock source

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -110,7 +110,11 @@ void config_plli2s(void)
 				       STM32_PLLI2S_N_MULTIPLIER,
 				       plli2sr(STM32_PLLI2S_R_DIVISOR));
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f412_plli2s_clock)
+#ifdef RCC_PLLR_I2S_CLKSOURCE_SUPPORT
 	LL_RCC_PLL_ConfigDomain_I2S(get_pll_source(),
+#else
+	LL_RCC_PLLI2S_ConfigDomain_I2S(get_pll_source(),
+#endif
```
STM2F411 should define its own compatible then.

FRASTM replied:
new st,stm32f411-plli2s-clock  added,

erwango commented on dts/arm/st/f4/stm32f411.dtsi:
```diff
@@ -6,7 +6,17 @@
 
 #include <st/f4/stm32f401.dtsi>
 
+/delete-node/ &plli2s;
```
Not required

FRASTM replied:
removed

erwango commented on dts/arm/st/f4/stm32f446.dtsi:
```diff
@@ -8,6 +8,8 @@
 #include <zephyr/dt-bindings/clock/stm32f410_clock.h>
 #include <zephyr/dt-bindings/memory-controller/stm32-fmc-sdram.h>
 
+/delete-node/ &plli2s;
+
```
Not required

FRASTM replied:
removed

erwango reviewed (changes requested):
In commit message of "drivers: clock_control: stm32 mco driver get define from DTS", please mention that DTS configuration has been introduced and Kconfig method deprecated two releases before and can be then safely removed.
Please also update migration guide on this.

FRASTM committed:
doc: Migration guide v4.1: STM32 addition for the MCO

STM32 configuration for the MCO clock source and prescaler is done
by the DTS and no more by Kconfig (removed).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -117,16 +111,13 @@ static uint32_t get_sysclk_frequency(void)
+	    (src_clk == STM32_SRC_PCLK1) || (src_clk == STM32_SRC_PCLK2) ||
 	    (src_clk == STM32_SRC_PCLK3) ||
 	    ((src_clk == STM32_SRC_HSE) && IS_ENABLED(STM32_HSE_ENABLED)) ||
 	    ((src_clk == STM32_SRC_HSI16) && IS_ENABLED(STM32_HSI_ENABLED)) ||
```
This change makes code harder to read and maintain.

FRASTM replied:
abandoned

erwango commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -103,10 +99,8 @@ static uint32_t get_pllout_frequency(uint32_t pllsrc_freq,
-					STM32_PLL_N_MULTIPLIER,
-					STM32_PLL_R_DIVISOR);
+	return get_pllout_frequency(get_pllsrc_frequency(PLL1_ID), STM32_PLL_M_DIVISOR,
+				    STM32_PLL_N_MULTIPLIER, STM32_PLL_R_DIVISOR);
```
This change makes code harder to read and maintain.

FRASTM replied:
abandoned

erwango commented on drivers/clock_control/clock_stm32_ll_u5.c:
```diff
@@ -18,23 +17,22 @@
 /* Macros to fill up prescaler values */
-#define z_ahb_prescaler(v) LL_RCC_SYSCLK_DIV_ ## v
-#define ahb_prescaler(v) z_ahb_prescaler(v)
+#define z_ahb_prescaler(v) LL_RCC_SYSCLK_DIV_##v
```
Are we sure this change is safe ?

FRASTM replied:
abandoned

erwango commented on samples/boards/st/mco/sample.yaml:
```diff
@@ -2,5 +2,6 @@ sample:
   sample.board.stm32.mco:
-    platform_allow: nucleo_u5a5zj_q stm32f746g_disco
+    platform_allow: nucleo_u5a5zj_q stm32f746g_disco nucleo_f411re
+    filter: not CONFIG_SOC_SERIES_STM32WB0X
```
Why specifically WB0 ? There are other series not enabling MCO.
Also, filter should not be required if `platform_allow` is used

erwango commented on samples/boards/st/mco/README.rst:
```diff
@@ -16,6 +16,12 @@ Requirements
 The SoC should support MCO functionality and use a pin that has the MCO alternate function.
 To support another board, add a dts overlay file in boards folder.
 Make sure that the output clock is enabled in dts overlay file.
+Depending on the stm32 serie, several clock source and prescaler are possible for each MCOx.
+The clock source is set by the DTS among the possible values for each stm32 serie.
+The prescaler is set by the DTS, through the property ``prescaler = <MCOx_PRE(n)>;``
+where ``n`` depends on the stm32 serie.
+
+It is required to check the Reference Manual to  configure the DTS correctly.
```
We should rather enhance binding documentation and redirect to it.

erwango commented on samples/boards/st/mco/boards/nucleo_f411re.overlay:
```diff
@@ -0,0 +1,32 @@
+	status = "okay";
+};
+
+#define MCO2_PRE_DIV_5 7
+
+&mco2 {
+	clocks = <&rcc STM32_SRC_PLLI2S_R MCO2_SEL(1)>;
+	prescaler = <MCO2_PRE(MCO2_PRE_DIV_5)>; /* set 0b111 in the MCO2PRE of the RCC CFGR */
```
Usage of `MCO2_PRE_DIV_5` greatly increase usability of the property.
I think we should add as part of `include/zephyr/dt-bindings/clock/stm32yy_clock.h` files directly so that all users could benefit.

FRASTM replied:
adding MCO2_PRE_DIV_n for stm32XX_clock.h bindings files

erwango commented on samples/boards/st/mco/src/main.c:
```diff
@@ -7,23 +7,29 @@
 #include <zephyr/drivers/pinctrl.h>
 
+#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(mco1)) || DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(mco2))
+const struct device *dev;
```
Use of a single const dev won't work if both MCO are enabled

erwango reviewed (changes requested):
I'd drop the clang commit entirely as most of the changes are bad from readbility/maintainability point of view.

FRASTM committed:
include: binding defines division factor for stm32 MCO prescaler

Depending on the stm32 serie the MCO1/2 prescaler is a value
set in the CFGR register to divide the MCO output clock.
Use the same model based on the RefMan for other stm32 series
than stm32C0/F4/F7/H5/H7, once the MCO is in the DTS.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on include/zephyr/dt-bindings/clock/stm32c0_clock.h:
```diff
@@ -81,4 +81,23 @@
 #define MCO2_PRE(val)           STM32_MCO_CFGR(val, 0x7, 20, CFGR1_REG)
 
+/* MCO prescaler : division factor */
+#define MCO1_PRE_DIV_1   0
```
Nice. Please also update https://github.com/zephyrproject-rtos/zephyr/blame/main/dts/bindings/clock/st%2Cstm32-clock-mco.yaml

FRASTM replied:
done

FRASTM committed:
drivers: clock control: stm32 pll clock config for I2S

The stm32F41x have a PLLI2S M divider for their PLL I2S
but others like the stm32F401 or stm32F74x have the PLL M
divider from the main PLL : might affect the sysclock.
LL Function is the same for configuring the PLL I2S but
parameter could depends on the stm32 serie.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -109,6 +109,11 @@ void config_plli2s(void)
 				       pllm(STM32_PLLI2S_M_DIVISOR),
 				       STM32_PLLI2S_N_MULTIPLIER,
 				       plli2sr(STM32_PLLI2S_R_DIVISOR));
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f411_plli2s_clock)
+	LL_RCC_PLLI2S_ConfigDomain_I2S(get_pll_source(),
+				       plli2sm(STM32_PLLI2S_M_DIVISOR),
+				       STM32_PLLI2S_N_MULTIPLIER,
+				       plli2sr(STM32_PLLI2S_R_DIVISOR));
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32f412_plli2s_clock)
 	LL_RCC_PLL_ConfigDomain_I2S(get_pll_source(),
```
right, Calling the LL_RCC_PLL_ConfigDomain_I2S(get_pll_source(),  for the stm32f412 compatible was wrong, the PLLI2S configuration is expected here. the  LL_RCC_PLLI2S_ConfigDomain_I2S is needed, with parameter divM, mul N, div R. 
Note that div M is either dedicated PLL I2S M divider for some stm32 devices (ex on stm32f41x)  or the PLL M divider  (ex on stm32f74x)

FRASTM commented:
add the MCO prescaler : division factor  as common to stm32 series, not defined yet all the stm32 series
Use the same LL_RCC_PLLI2S_ConfigDomain_I2S to configure the PLL I2S, only the div M factor changes depending on the stm32 serie

erwango commented on include/zephyr/dt-bindings/clock/stm32f7_clock.h:
```diff
@@ -82,6 +82,20 @@
 #define MCO2_SEL(val)           STM32_MCO_CFGR(val, 0x3, 30, CFGR_REG)
 #define MCO2_PRE(val)           STM32_MCO_CFGR(val, 0x7, 27, CFGR_REG)
+
+/* MCO prescaler : division factor */
+#define MCO1_PRE_DIV_1 0
+#define MCO1_PRE_DIV_2 4
+#define MCO1_PRE_DIV_3 5
+#define MCO1_PRE_DIV_4 6
+#define MCO1_PRE_DIV_5 7
+
+#define MCO2_PRE_DIV_1 0
+#define MCO2_PRE_DIV_2 4
+#define MCO2_PRE_DIV_3 5
+#define MCO2_PRE_DIV_4 6
+#define MCO2_PRE_DIV_5 7
```
Since value is always the same between MCO 1 and MCO 2, lets provide only one set of definitions for each series.
```
#define MCO_PRE_DIV_1 0
#define MCO_PRE_DIV_2 4
#define MCO_PRE_DIV_3 5
#define MCO_PRE_DIV_4 6
#define MCO_PRE_DIV_5 7
```

FRASTM commented on samples/boards/st/mco/boards/nucleo_f411re.overlay:
```diff
@@ -0,0 +1,30 @@
+
+&mco2 {
+	clocks = <&rcc STM32_SRC_PLLI2S_R MCO2_SEL(1)>;
+	prescaler = <MCO2_PRE(MCO2_PRE_DIV_5)>; /* set 0b111 in the MCO2PRE of the RCC CFGR */
```
comment removed

FRASTM commented on samples/boards/st/mco/boards/nucleo_f411re.overlay:
```diff
@@ -0,0 +1,30 @@
+/*	clocks = <&rcc STM32_SRC_LSE MCO1_SEL(1)>; */
+/*	clocks = <&rcc STM32_SRC_HSE MCO1_SEL(2)>; */
+/*	clocks = <&rcc STM32_SRC_PLL_P MCO1_SEL(3)>; */
+	prescaler = <MCO1_PRE(4)>; /* set 0b100 in the MCO1PRE of the RCC CFGR */
```
with <MCO1_PRE(MCO_PRE_DIV_2)>;

erwango commented on drivers/clock_control/Kconfig.stm32:
```diff
@@ -97,310 +97,6 @@ config CLOCK_STM32_MCO
-	range 1 15 if SOC_SERIES_STM32H7X || SOC_SERIES_STM32H7RSX || SOC_SERIES_STM32H5X
-	help
-	  Prescaler for MCO2 output clock
+	# MCO configuration from DT-based for compiling the MCO driver.
```
```
	help
	  Allows to output various different clock sources onto the MCO pin using a configurable prescaler.
```

FRASTM replied:
changed

erwango commented on samples/boards/st/mco/README.rst:
```diff
@@ -16,6 +16,13 @@ Requirements
+The clock source is set by the DTS among the possible values for each stm32 serie.
+The prescaler is set by the DTS, through the property ``prescaler = <MCOx_PRE(MCO_PRE_DIV_n)>;``
+
+See the ``dts/bindings/clock/st,stm32-clock-mco.yaml``.
```
```suggestion
See :zephyr_file:`dts/bindings/clock/st,stm32-clock-mco.yaml`
```

FRASTM replied:
changed

erwango reviewed (changes requested):
Otherwise LGTM

erwango commented:
@FRASTM Please rebase to fix CI issue

FRASTM commented:
rebased on 065bd322200ccf2f9d717fd258ec966a522ce733

FRASTM commented:
rebase on  af68d97507188249e2f9677119fc716a4d04989e

erwango commented:
@nordic-krch PTAL

fabiobaltieri commented:
All changes are in ST specific file, reassigning.

fabiobaltieri commented:
```
Area matches: {<Area STM32 Platforms>: 20, <Area Drivers: Clock control>: 8, <Area Release Notes>: 1, <Area Samples>: 0}
labels: {'area: Clock control', 'platform: STM32', 'area: Samples', 'Release Notes'}
collab: ['erwango', 'FRASTM', 'gautierg-st', 'GeorgeCGV', 'marwaiehm-st', 'mathieuchopstm', 'nordic-krch', 'fabiobaltieri', 'kartben', 'nashif']
Submitted by: FRASTM
candidate maintainers: {'erwango': 20, 'nordic-krch': 8, 'fabiobaltieri': 1, 'kartben': 1}
Picked assignees: ['nordic-krch'] (27.59% ownership)
```

shrugs
